File: webflux-dialogue/src/main/java/com/study/webflux/rag/application/monitoring/context/PipelineContext.java
package com.study.webflux.rag.application.monitoring.context;

import com.study.webflux.rag.application.monitoring.monitor.DialoguePipelineTracker;
import reactor.util.context.Context;
import reactor.util.context.ContextView;

public final class PipelineContext {

	public static final Object TRACKER_KEY = new Object();

	private PipelineContext() {
	}

	public static Context withTracker(Context context, DialoguePipelineTracker tracker) {
		return context.put(TRACKER_KEY, tracker);
	}

	public static DialoguePipelineTracker requireTracker(ContextView contextView) {
		return contextView.get(TRACKER_KEY);
	}

	public static DialoguePipelineTracker findTracker(ContextView contextView) {
		if (contextView.hasKey(TRACKER_KEY)) {
			return contextView.get(TRACKER_KEY);
		}
		return null;
	}
}

File: application/src/main/kotlin/com/miyou/app/application/monitoring/context/PipelineContext.kt
package com.miyou.app.application.monitoring.context

import com.miyou.app.application.monitoring.monitor.DialoguePipelineTracker
import reactor.util.context.Context
import reactor.util.context.ContextView

object PipelineContext {
    val TRACKER_KEY: Any = Any()

    fun withTracker(
        context: Context,
        tracker: DialoguePipelineTracker,
    ): Context = context.put(TRACKER_KEY, tracker)

    fun requireTracker(contextView: ContextView): DialoguePipelineTracker = contextView.get(TRACKER_KEY)

    fun findTracker(contextView: ContextView): DialoguePipelineTracker? =
        if (contextView.hasKey(TRACKER_KEY)) {
            contextView.get(TRACKER_KEY)
        } else {
            null
        }
}
